proc/msg-nommu.c

```diff
@@ -129,19 +145,50 @@ int proc_recv(u32 port, msg_t *msg, msg_rid_t *rid)
 	}
 
 	if (proc_portRidAlloc(p, kmsg) < 0) {
-		hal_spinlockSet(&p->spinlock, &sc);
-		kmsg->state = msg_rejected;
-		proc_threadWakeup(&kmsg->threads);
-		hal_spinlockClear(&p->spinlock, &sc);
-
-		port_put(p, 0);
+		proc_msgReject(kmsg, p);
 		return -ENOMEM;
 	}
 
 	*rid = lib_idtreeId(&kmsg->idlinkage);
 
 	hal_memcpy(msg, kmsg->msg, sizeof(*msg));
 
+	kmsg->imapped = NULL;
+	kmsg->omapped = NULL;
+
+	if ((kmsg->msg->i.data != NULL) && (kmsg->msg->i.size != 0) && (current->process != NULL) &&
+			(pmap_isAllowed(current->process->pmapp, kmsg->msg->i.data, kmsg->msg->i.size) == 0)) {
+		idata = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->i.size),
+			PROT_READ | PROT_USER, NULL, -1, MAP_ANONYMOUS);
+		if (idata == NULL) {
+			/* Free RID */
+			(void)proc_portRidGet(p, *rid);
+			proc_msgReject(kmsg, p);
+			return -ENOMEM;
+		}
+		hal_memcpy(idata, kmsg->msg->i.data, kmsg->msg->i.size);
+		kmsg->imapped = idata;
+		msg->i.data = idata;
+	}
+
+	if ((kmsg->msg->o.data != NULL) && (kmsg->msg->o.size != 0) && (current->process != NULL) &&
+			(pmap_isAllowed(current->process->pmapp, kmsg->msg->o.data, kmsg->msg->o.size) == 0)) {
+		msg->o.data = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->o.size),
```

Comment:
I'd add some space here
```suggestion

		msg->o.data = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->o.size),
```

Reply:
Done